py-rse/errors.Rmd

```diff
@@ -155,87 +154,134 @@ except ArithmeticError as error:
+This least to the rule "throw low, catch high":
 write most of your code without exception handlers,
 since there's nothing useful you can do in the middle of a small utility function,
 but put a few handlers in the main body of your program
 to catch and report all errors.
 
 ## What exceptions should I raise and catch? {#py-rse-errors-where}
 
-Broadly speaking there are two approaches to error handling:
-"look before your leap"
-and
-"it's easier to ask for forgiveness than permission".
-When using the first,
-we check if an operation will be allowed before trying it:
+Now that we have learned about error handling,
+we can go ahead and add it to our zipf's law code.
+First,
+you will notice that there is already some error handling built in,
+e.g. if you try to read a file that does not exist,
+the error handling inside the `open` function throws a `FileNotFoundError`:
 
-```python
-LOG_DIR = "/tmp/mylogdir"
+```shell
+python bin/collate.py results/no-file
+```
 
-def write_message(filename, message):
-    if os.path.isdir(LOG_DIR):
-        path = os.path.join(LOG_DIR, filename)
-        with open(path, 'a') as writer:
-            writer.write(message)
-    else:
-        print('No log directory {}'.format(LOG_DIR), file=sys.stderr)
+```text
+Traceback (most recent call last):
+  File "bin/collate.py", line 27, in <module>
+    main(args)
+  File "bin/collate.py", line 17, in main
+    with open(file_name, 'r') as reader:
+FileNotFoundError: [Errno 2] No such file or directory: 'results/none.csv'
 ```
 
-With the second,
-we try the operation and catch the exception if one is raised:
+But what happens if we are trying to read an file that exists,
+but was not created by `countwords.py`?
 
-```python
-LOG_DIR = "/tmp/mylogdir"
 
-def write_message(filename, message):
-    try:
-        path = os.path.join(LOG_DIR, filename)
-        with open(path, 'a') as writer:
-            writer.write(message)
-    except OSError as error:
-        print('Unable to write log message to {}: {}'.format(path, error))
-```{: title="logging/ask_permission_not_forgiveness.py"}
-
-The first approach doesn't actually need exception handlers,
-except it does:
-programmers can never anticipate all of the things that can go wrong in the real world,
-so there should always be an `except` somewhere.
-Since that's the case,
-most programmers prefer the second approach.
-
-When we are raising exceptions ourselves,
-we should make the information as specific as possible.
-For example,
-suppose a program contains this:
+```shell
+$ python bin/collate.py Makefile
+```
+
+```text
+Traceback (most recent call last):
+  File "bin/collate.py", line 27, in <module>
+    main(args)
+  File "bin/collate.py", line 18, in main
+    update_counts(reader, word_counts)
+  File "bin/collate.py", line 10, in update_counts
+    for word, count in csv.reader(reader):
+ValueError: not enough values to unpack (expected 2, got 1)
+```
+
+This error is not that easy to understand.
```

Comment:
"This error is not easy for someone to understand if they're not familiar with the code's internals."